vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdIT.java

```diff
@@ -0,0 +1,107 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.component.grid.it;
+
+import com.vaadin.flow.component.checkbox.testbench.CheckboxElement;
+import com.vaadin.flow.component.grid.testbench.GridElement;
+import com.vaadin.flow.testutil.TestPath;
+import com.vaadin.tests.AbstractComponentIT;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+@TestPath("vaadin-grid/select-item-with-identical-id")
+public class SelectItemWithIdenticalIdIT extends AbstractComponentIT {
+
+    private WebElement addGridButton;
+
+    private WebElement selectAnotherItemButton;
+
+    private WebElement deselectItemButton;
+
+    private CheckboxElement useMultiSelectCheckbox;
+
+    @Before
+    public void init() {
+        open();
+        addGridButton = findElement(By.id("add-grid-button"));
+        selectAnotherItemButton = findElement(
+                By.id("select-another-item-button"));
+        deselectItemButton = findElement(By.id("deselect-item-button"));
+        useMultiSelectCheckbox = $(CheckboxElement.class).first();
+    }
+
+    @Test
+    public void addGridWithSelection_itemIsNotUpdated() {
+        addGridButton.click();
+        GridElement grid = $(GridElement.class).waitForFirst();
+
+        Assert.assertEquals("1", grid.getCell(0, 0).getText());
+    }
+
+    @Test
+    public void addGridWithSelection_deselectItem_itemIsNotUpdated() {
+        addGridButton.click();
+        GridElement grid = $(GridElement.class).waitForFirst();
+
+        deselectItemButton.click();
+
+        Assert.assertEquals("1", grid.getCell(0, 0).getText());
+    }
+
+    @Test
+    public void addGridWithSelection_selectAnotherItem_itemIsNotUpdated() {
+        addGridButton.click();
+        GridElement grid = $(GridElement.class).waitForFirst();
+
+        selectAnotherItemButton.click();
+
+        Assert.assertEquals("2", grid.getCell(1, 0).getText());
+    }
+
+    @Test
+    public void setMultiSelect_addGridWithSelection_itemIsNotUpdated() {
+        useMultiSelectCheckbox.setChecked(true);
+        addGridButton.click();
+        GridElement grid = $(GridElement.class).waitForFirst();
+
+        Assert.assertEquals("1", grid.getCell(0, 1).getText());
+    }
+
+    @Test
+    public void setMultiSelect_addGridWithSelection_deselectItem_itemIsNotUpdated() {
+        useMultiSelectCheckbox.setChecked(true);
+        addGridButton.click();
+        GridElement grid = $(GridElement.class).waitForFirst();
+
+        deselectItemButton.click();
+
+        Assert.assertEquals("1", grid.getCell(0, 1).getText());
+    }
+
+    @Test
+    public void setMultiSelect_addGridWithSelection_selectAnotherItem_itemIsNotUpdated() {
+        useMultiSelectCheckbox.setChecked(true);
```

Comment:
Is it essential to set the multi-select mode before adding the grid to the DOM?

Reply:
The selected item is cached in the selection model. Therefore, the order actually plays a role in the end result. For example, calling `refreshAll` afterwards loses the selection that was made before the selection mode change.